framework/src/vectorpostprocessors/MaterialVectorPostprocessor.C

```diff
@@ -0,0 +1,115 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "MaterialVectorPostprocessor.h"
+#include "Material.h"
+#include "IndirectSort.h"
+
+#include "libmesh/quadrature.h"
+
+#include <numeric>
+
+template<>
+InputParameters validParams<MaterialVectorPostprocessor>()
+{
+  InputParameters params = validParams<ElementVectorPostprocessor>();
+  params.addRequiredParam<std::string>("material", "Material for which all properties will be recorded.");
+  params.addRequiredParam<std::vector<unsigned int>>("elem_ids", "Element IDs to print data for (others are ignored).");
+  return params;
+}
+
+MaterialVectorPostprocessor::MaterialVectorPostprocessor(const InputParameters & parameters) :
+    ElementVectorPostprocessor(parameters),
+    _elem_filter(getParam<std::vector<unsigned int>>("elem_ids").begin(), getParam<std::vector<unsigned int>>("elem_ids").end()),
+    _elem_ids(declareVector("elem_id")),
+    _qp_ids(declareVector("qp_id"))
+{
+  auto & mat = getMaterialByName(getParam<std::string>("material"), true);
+  auto & prop_names = mat.getSuppliedItems();
+  for (auto & prop : prop_names)
+  {
+    _prop_vecs.push_back(&declareVector(prop));
+    _prop_refs.push_back(&getMaterialProperty<Real>(prop));
+  }
+}
+
+void
+MaterialVectorPostprocessor::execute()
+{
+  unsigned int elem_id = _current_elem->id();
+  if (_elem_filter.count(elem_id) == 0)
+    return;
+
+  unsigned int nqp = _qrule->n_points();
+  for (unsigned int qp = 0; qp < nqp; qp++)
+  {
+    _elem_ids.push_back(elem_id);
+    _qp_ids.push_back(qp);
+  }
+
+  for (unsigned int i = 0; i < _prop_vecs.size(); i++)
+  {
+    auto prop = _prop_vecs[i];
+    auto vals = _prop_refs[i];
+    for (unsigned int qp = 0; qp < nqp; qp++)
+      prop->push_back((*vals)[qp]);
```

Comment:
Because we can't know a-priori in the code for this postprocessor what the c++ type of the material property the user will have linked to, we can't guard against the MOOSE_ABORT call in the ``getMaterialProperty<[cpptype]>(...)`` call.  The only way to support other/multiple types is to have separate postprocessors for each class type.  We can template a class making it easier to create instantiations for different types, but the logic for their handling will be different enough that it really wouldn't buy you anything.